### ISSUE TITLE
getting scale and offset for a band returns default values instead of correct values

### ISSUE THREAD

**Issue by anielsen001** [yeesian/ArchGDAL.jl#183]
I'm trying to read the scale and offset for a band, using the snippet of code below. I can use python osgeo.gdal and correctly access these values. It looks like, this issue affects GDAL.jl since the ArchGDAL code calls a similar function in GDAL.jl. So this issue might just be more of an informational type, and I have not tried the same example with the GDAL.jl interface yet.

```
using ArchGDAL

# https://geographiclib.sourceforge.io/html/geoid.html
filename = "/nasdata/egm96/geoids/egm96-5.pgm"

dataset = ArchGDAL.readraster(filename)
band = ArchGDAL.getband(dataset, 1)
scale = ArchGDAL.getscale(band) # = 1.0 should be 0.003
offset = ArchGDAL.getoffset(band) # = 0.0 should be -108
````

The metadata file is correctly found by GDAL

```
ulia> dataset = ArchGDAL.readraster(filename)
GDAL Dataset (Driver: PNM/Portable Pixmap Format (netpbm))
File(s): 
  /nasdata/egm96/geoids/egm96-5.pgm
  /nasdata/egm96/geoids/egm96-5.pgm.aux.xml

Dataset (width x height): 4320 x 2161 (pixels)
Number of raster bands: 1
  [GA_ReadOnly] Band 1 (Gray): 4320 x 2161 (UInt16)
```

and contains the following 
```
shell> cat /nasdata/egm96/geoids/egm96-5.pgm.aux.xml
<PAMDataset>
  <SRS>GEOGCS[&quot;WGS 84&quot;,DATUM[&quot;WGS_1984&quot;,SPHEROID[&quot;WGS 84&quot;,6378137,298.257223563,AUTHORITY[&quot;EPSG&quot;,&quot;7030&quot;]],AUTHORITY[&quot;EPSG&quot;,&quot;6326&quot;]],PRIMEM[&quot;Greenwich&quot;,0],UNIT[&quot;degree&quot;,0.017453292519943295769],AUTHORITY[&quot;EPSG&quot;,&quot;4326&quot;]]</SRS>
  <GeoTransform>-0.04166666666666666,0.08333333333333333,0,90.04166666666666666,0,-0.08333333333333333</GeoTransform>
  <Metadata>
    <MDI key="Description">WGS84 EGM96, 5-minute grid</MDI>
    <MDI key="URL">http://earth-info.nga.mil/GandG/wgs84/gravitymod/egm96/egm96.html</MDI>
    <MDI key="DateTime">2009-08-29 18:45:03</MDI>
    <MDI key="MaxBilinearError">0.140</MDI>
    <MDI key="RMSBilinearError">0.005</MDI>
    <MDI key="MaxCubicError">0.003</MDI>
    <MDI key="RMSCubicError">0.001</MDI>
    <MDI key="Offset">-108</MDI>
    <MDI key="Scale">0.003</MDI>
    <MDI key="AREA_OR_POINT">Point</MDI>
    <MDI key="Vertical_Datum">WGS84</MDI>
    <MDI key="Tie_Point_Location">pixel_corner</MDI>
  </Metadata>
</PAMDataset>
```

**Comment by visr** [comment]
Hmm I can reproduce. Not sure what is going on here. Could you also share the Python code you used?

The wrapping code around this function is very thin, so you'd think the GDAL build we use may be to blame? Though if I convert to GeoTIFF first with my Python build, I get the same thing.

Perhaps it's worth checking the `pbSuccess` flag value? Not many other ideas.
https://github.com/JuliaGeo/GDAL.jl/blob/d44dc737ec16a60d1edb81bb71e0cb5f0666f165/src/gdal_h.jl#L2038-L2046

Isn't this somehow just arbitrary metadata and not in the way that GDAL expects scale and offset? This shows the metadata, and with this (or rather `metadataitem`) the value can also be retrieved:

```julia
julia> dataset = ArchGDAL.read(filename)
GDAL Dataset (Driver: PNM/Portable Pixmap Format (netpbm))
File(s):
  c:\Users\visser_mn\Downloads\egm96-5\geoids\egm96-5.pgm
  c:\Users\visser_mn\Downloads\egm96-5\geoids\egm96-5.pgm.aux.xml

Dataset (width x height): 4320 x 2161 (pixels)
Number of raster bands: 1
  [GA_ReadOnly] Band 1 (Gray): 4320 x 2161 (UInt16)


julia> AG.metadata(dataset)
12-element Vector{String}:
 "AREA_OR_POINT=Point"
 "DateTime=2009-08-29 18:45:03"
 "Description=WGS84 EGM96, 5-minute grid"
 "MaxBilinearError=0.140"
 "MaxCubicError=0.003"
 "Offset=-108"
 "RMSBilinearError=0.005"
 "RMSCubicError=0.001"
 "Scale=0.003"
 "Tie_Point_Location=pixel_corner"
 "URL=http://earth-info.nga.mil/GandG/wgs84/gravitymod/egm96/egm96.html"
 "Vertical_Datum=WGS84"
```

I found that I had to do `read` instead of `readraster` for `metadata` to work, since `RasterDataset` does not have a `ptr` field, whereas the method expects that. So that should be fixed, we can add method definitions for this.

https://github.com/yeesian/ArchGDAL.jl/blob/1dc8ca8360f2e6a2ee2319fd63c3ba973034f8a6/src/raster/array.jl#L27-L30

https://github.com/yeesian/ArchGDAL.jl/blob/1dc8ca8360f2e6a2ee2319fd63c3ba973034f8a6/src/utils.jl#L70-L71

**Comment by anielsen001** [comment]
I'm happy to share my python code, I had it wrapped in a class and simplified it for discussion purposes, but it works
```
from osgeo import gdal, gdal_array
gdal.UseExceptions()

filename = "/nasdata/egm96/geoids/egm96-5.pgm"
raster = gdal.Open(filename) 
meta = raster.GetMetadata()
scale = float(meta['Scale'])
offset = float(meta['Offset'])
```
I wrote this construct a while ago and realize as I look at it, that it's not quite the same idea as a simple accessor method for scale and offset. To be honest, I don't know how GDAL expects it, so maybe it is the data file that's got an issue. 

The `meta` dictionary that I read from GDAL in python is
```
In [5]: meta                                                                                                                                                                                            
Out[5]: 
{'AREA_OR_POINT': 'Point',
 'DateTime': '2009-08-29 18:45:03',
 'Description': 'WGS84 EGM96, 5-minute grid',
 'MaxBilinearError': '0.140',
 'MaxCubicError': '0.003',
 'Offset': '-108',
 'RMSBilinearError': '0.005',
 'RMSCubicError': '0.001',
 'Scale': '0.003',
 'Tie_Point_Location': 'pixel_corner',
 'URL': 'http://earth-info.nga.mil/GandG/wgs84/gravitymod/egm96/egm96.html',
 'Vertical_Datum': 'WGS84'}
```

Thank you for including the `read` vs `readraster`, becuase I could not figure out how to even get the meta data the way I was accessing the raster. If I can at least get the metadata I can make it work, the way I've done it with my python code.

**Comment by anielsen001** [comment]
Thanks to your suggestion about `read` vs. `readrraster`, I was able to achieve the same result with this bit of juila code

```
using ArchGDAL

egmfilename = "/nasdata/egm96/geoids/egm96-5.pgm"
ds = ArchGDAL.read(egmfilename)
egm96 = ArchGDAL.read(ds,1) # will be UInt16 in Hex
egm96 = convert(Matrix{Float64},egm96) # Convert to float 64

meta = ArchGDAL.metadata(ds)

d = Dict()
for m in meta
    k,v=split(m,"=")
    d[k]=v
end

emg96 = parse(Float64,d["Scale"]) .* egm96 .+ parse(Float64,d["Offset"])
```

A minor difference being that the python code returns a dictionary of metadata, while the juila version returns a vector of strings that I parsed into a dictionary.

the Julia code also returns the UInt16 values as Hex, so I convert to Float64 in order to apply the scale and offsets.

**Comment by visr** [comment]
Here is a somewhat condensed version, using `metadataitem`:
```julia
using ArchGDAL

egmfilename = "/nasdata/egm96/geoids/egm96-5.pgm"

ds = ArchGDAL.read(egmfilename)
egm96_unscaled = ArchGDAL.read(ds,1)

meta = ArchGDAL.metadata(ds)
scale = parse(Float64, ArchGDAL.metadataitem(ds, "Scale"))
offset = parse(Float64, ArchGDAL.metadataitem(ds, "Offset"))

emg96 = egm96_unscaled .* scale .+ offset
```

This will also work with `readraster` after #186 is merged.

> A minor difference being that the python code returns a dictionary of metadata, while the juila version returns a vector of strings that I parsed into a dictionary.

Feel free to create a separate issue to discuss changing this. Now would be a good moment, since with #179 there is a clear push towards a 1.0 version with some other breaking changes.

> the Julia code also returns the UInt16 values as Hex

UInt16 values are indeed by default shown as hexadecimal numbers, but that doesn't mean you cannot use them like you would with Int64 integers for instance. The types also promote automatically, see for instance:

```julia
julia> UInt16(2)
0x0002

julia> UInt16(2) * 3.0
6.0
```

Since in your Python example you are also just using the metadata, and not a wrapper around `GDALGetRasterScale` like `ArchGDAL.getscale`, I'll close this issue, since it doesn't demonstrate an issue.
